# LABORATORIO 1: INTEGRAZIONE, PULIZIA, E OPERAZIONI SUI DATI

a.a. 2024/2025, 06/03/2025

In questo laboratorio impareremo a trattare dati provenienti da sorgenti diverse, cercando di ottenere informazioni integrate e complete.
Come esempio, ci concentreremo su dati riguardanti film e serie tv dei maggiori servizi di streaming.

In [1]:
import pandas as pd

### Carico i dati di Netflix

In [2]:
netflix_titles = pd.read_csv("netflix_titles.csv")

Diamo un'occhiata a come è organizzato il dataset

In [3]:
netflix_titles.head()

Unnamed: 0        id                                title   type  \
0           0  ts300399  Five Came Back: The Reference Films   SHOW   
1           1   tm82169                                Rocky  MOVIE   
2           2   tm17823                               Grease  MOVIE   
3           3   tm69975                             Rocky II  MOVIE   
4           4  tm127384      Monty Python and the Holy Grail  MOVIE   

   release_year age_certification  runtime                 genres  \
0          1945             TV-MA       51      ['documentation']   
1          1976                PG      119     ['drama', 'sport']   
2          1978                PG      110  ['comedy', 'romance']   
3          1979                PG      119     ['drama', 'sport']   
4          1975                PG       91  ['fantasy', 'comedy']   

  production_countries  seasons  ... tmdb_popularity  tmdb_score  \
0               ['US']      1.0  ...           0.601         NaN   
1               ['US']      NaN  ...          99.713       7.783   
2               ['US']      NaN  ...          33.160       7.406   
3               ['US']      NaN  ...          74.191       7.200   
4               ['GB']      NaN  ...          20.964       7.804   

                     director  \
0                         NaN   
1            John G. Avildsen   
2              Randal Kleiser   
3          Sylvester Stallone   
4  Terry Gilliam, Terry Jones   

                                                cast         country  \
0                                                NaN   United States   
1  Sylvester Stallone, Talia Shire, Burt Young, C...   United States   
2  John Travolta, Olivia Newton-John, Stockard Ch...   United States   
3  Sylvester Stallone, Talia Shire, Burt Young, C...   United States   
4  Graham Chapman, John Cleese, Eric Idle, Terry ...  United Kingdom   

         date_added rating  duration  \
0  February 9, 2021  TV-MA  1 Season   
1    August 1, 2019     PG   120 min   
2  November 1, 2019     PG   110 min   
3    August 1, 2019     PG   119 min   
4   January 2, 2019     PG    92 min   

                                      listed_in  \
0                                      TV Shows   
1    Action & Adventure, Classic Movies, Dramas   
2    Classic Movies, Comedies, Music & Musicals   
3                         Dramas, Sports Movies   
4  Action & Adventure, Classic Movies, Comedies   

                                         description  
0  This collection includes 12 World War II-era p...  
1  Sylvester Stallone shot to fame as Rocky Balbo...  
2  John Travolta and Olivia Newton-John star in t...  
3  Featuring a rousing climax, this engaging sequ...  
4  The Monty Python comedy clan skewers King Arth...  

[5 rows x 23 columns]

ed allo schema (notate che con head() alcune delle colonne, per problemi di spazio, non compaiono)

In [192]:
netflix_titles.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3645 entries, 0 to 3644
Data columns (total 23 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Unnamed: 0            3645 non-null   int64  
 1   id                    3645 non-null   object 
 2   title                 3645 non-null   object 
 3   type                  3645 non-null   object 
 4   release_year          3645 non-null   int64  
 5   age_certification     2003 non-null   object 
 6   runtime               3645 non-null   int64  
 7   genres                3645 non-null   object 
 8   production_countries  3645 non-null   object 
 9   seasons               1453 non-null   float64
 10  imdb_id               3417 non-null   object 
 11  imdb_score            3378 non-null   float64
 12  imdb_votes            3374 non-null   float64
 13  tmdb_popularity       3612 non-null   float64
 14  tmdb_score            3529 non-null   float64
 15  director             

### Carico i dati di Disney+

In [193]:
disney_titles = pd.read_csv("disney_titles.csv")

In [194]:
disney_titles.head()

Unnamed: 0        id                                   title show_type  \
0           0   tm61729  The Adventures of Ichabod and Mr. Toad     MOVIE   
1           1   tm61052                              Cinderella     MOVIE   
2           2  tm190870                              Cinderella     MOVIE   
3           3   tm87946                                   Dumbo     MOVIE   
4           4  tm207119                                   Dumbo     MOVIE   

   release_year age_certification  runtime  \
0          1950                 G       68   
1          1950                 G       74   
2          1997                 G       88   
3          1941                 G       64   
4          2019                PG      112   

                                              genres production_countries  \
0  ['horror', 'fantasy', 'animation', 'comedy', '...               ['US']   
1      ['fantasy', 'animation', 'family', 'romance']               ['US']   
2                   ['romance', 'family', 'fantasy']               ['US']   
3        ['animation', 'drama', 'family', 'fantasy']               ['US']   
4                    ['family', 'fantasy', 'action']         ['GB', 'US']   

   seasons    imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  \
0      NaN  tt0041094         6.9     16497.0           20.116       6.500   
1      NaN  tt0042332         7.3    164222.0           93.547       7.035   
2      NaN  tt0128996         6.7      8944.0            9.493       6.400   
3      NaN  tt0033563         7.2    134997.0           54.265       6.999   
4      NaN  tt3861390         6.3     73010.0           23.271       6.661   

                                          director  \
0         Jack Kinney, Clyde Geronimi, James Algar   
1  Wilfred Jackson, Hamilton Luske, Clyde Geronimi   
2  Wilfred Jackson, Hamilton Luske, Clyde Geronimi   
3                                       Tim Burton   
4                                       Tim Burton   

                                            country               from rating  \
0                                     United States  November 12, 2019      G   
1                                     United States  November 12, 2019      G   
2                                     United States  November 12, 2019      G   
3  United States, United Kingdom, Australia, Canada  November 12, 2019     PG   
4  United States, United Kingdom, Australia, Canada  November 12, 2019     PG   

  duration  
0   69 min  
1   77 min  
2   77 min  
3  117 min  
4  117 min

In [195]:
disney_titles.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1070 entries, 0 to 1069
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Unnamed: 0            1070 non-null   int64  
 1   id                    1070 non-null   object 
 2   title                 1070 non-null   object 
 3   show_type             1070 non-null   object 
 4   release_year          1070 non-null   int64  
 5   age_certification     854 non-null    object 
 6   runtime               1070 non-null   int64  
 7   genres                1070 non-null   object 
 8   production_countries  1070 non-null   object 
 9   seasons               263 non-null    float64
 10  imdb_id               807 non-null    object 
 11  imdb_score            792 non-null    float64
 12  imdb_votes            791 non-null    float64
 13  tmdb_popularity       1067 non-null   float64
 14  tmdb_score            1027 non-null   float64
 15  director             

Come abbiamo visto, i due dataset hanno molte colonne in comune, ma anche alcune differenze.

Proviamo comunque ad integrare le due tabelle, creandone una terza che metta insieme i dati della prima e quelli della seconda.

**tip**
Dovete "concatenare" i due dataFrame di pandas.

Provate a cercare la funzione corretta e scrivete il codice necessario.

In [196]:
# ... completare qui ...

integrate_titles = pd.concat([netflix_titles,disney_titles])

Adesso, stampiamo un po' di informazioni sul dataset integrato (come ad esempio le dimensioni) e diamo un'occhiata allo schema.

In [197]:
# ... completare qui ...

print("Dimensioni della tabella netflix_titles: ", netflix_titles.shape)
print("Dimensioni della tabella disney_titles: ", disney_titles.shape)
print("Dimensioni della tabella integrata: ",integrate_titles.shape)

integrate_titles.info()

Dimensioni della tabella netflix_titles:  (3645, 23)
Dimensioni della tabella disney_titles:  (1070, 20)
Dimensioni della tabella integrata:  (4715, 25)
<class 'pandas.core.frame.DataFrame'>
Index: 4715 entries, 0 to 1069
Data columns (total 25 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Unnamed: 0            4715 non-null   int64  
 1   id                    4715 non-null   object 
 2   title                 4715 non-null   object 
 3   type                  3645 non-null   object 
 4   release_year          4715 non-null   int64  
 5   age_certification     2857 non-null   object 
 6   runtime               4715 non-null   int64  
 7   genres                4715 non-null   object 
 8   production_countries  4715 non-null   object 
 9   seasons               1716 non-null   float64
 10  imdb_id               4224 non-null   object 
 11  imdb_score            4170 non-null   float64
 12  imdb_votes            4165

Nonostante le differenze tra le due tabelle, la funzione concat concatena le due tabelle, mettendo a Null i valori delle colonne presenti solo in una delle due ed in corrispondenza delle righe provenienti dal dataset in cui la colonna non è presente. Proviamo a verificare la presenza di Null

In [198]:
# ... completare qui ...

integrate_titles.isnull().sum()

Unnamed: 0                 0
id                         0
title                      0
type                    1070
release_year               0
age_certification       1858
runtime                    0
genres                     0
production_countries       0
seasons                 2999
imdb_id                  491
imdb_score               545
imdb_votes               550
tmdb_popularity           36
tmdb_score               159
director                1695
cast                    1435
country                  505
date_added              1070
rating                     2
duration                   1
listed_in               1070
description             1070
show_type               3645
from                    3646
dtype: int64

Ad esempio la colonna **listed_in**, che è presente solo in netflix_titles, ha 1070 valori Null, che corrisponde al numero di righe della tabella disney_titles.

Decidiamo di eliminare eventuali colonne non comuni. Possiamo farlo **prima** dell'integrazione, oppure direttamente sulla tabella concatenata.

**Tip**: confrontate i due schemi per capire quali siano le colonne non comuni. Attenzione!!! Potrebbero anche esserci colonne che sembrano non comuni, ma in realtà contengono la stessa tipologia di informazione e hanno solo nomi diversi!!

In [199]:
# ... completare qui ...

# Nota:
# listed_in è su netflix ma non disney
# description è su netflix ma non disney
# cast è su netflix ma non disney
# type e show_type sono la stessa cosa
# date_added in netflix e from in Disney sono la stessa cosa

integrate_titles=integrate_titles.drop(columns=["listed_in"], axis=1)
integrate_titles=integrate_titles.drop(columns=["description"], axis=1)
integrate_titles=integrate_titles.drop(columns=["cast"], axis=1)

integrate_titles.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4715 entries, 0 to 1069
Data columns (total 22 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Unnamed: 0            4715 non-null   int64  
 1   id                    4715 non-null   object 
 2   title                 4715 non-null   object 
 3   type                  3645 non-null   object 
 4   release_year          4715 non-null   int64  
 5   age_certification     2857 non-null   object 
 6   runtime               4715 non-null   int64  
 7   genres                4715 non-null   object 
 8   production_countries  4715 non-null   object 
 9   seasons               1716 non-null   float64
 10  imdb_id               4224 non-null   object 
 11  imdb_score            4170 non-null   float64
 12  imdb_votes            4165 non-null   float64
 13  tmdb_popularity       4679 non-null   float64
 14  tmdb_score            4556 non-null   float64
 15  director              3020

Andiamo ora alla ricerca di eventuali colonne che contengono la stessa informazione. Guardate in particolare (type, show_type),  (contry, production_countries), (date_added, from)

In [200]:
# ...completare qui ...

# dato che type e show_type sono due colonne che contengono la stessa informazione e hanno solo nome diverso,
# dobbiamo integrate le due colonne in una singola, buttando via l'altra
# Copiare la porzione non nulla di show_type dentro a type

# Anche date_added e from sono la stessa cosa, vale la cosa di prima
import numpy as np

integrate_titles.iloc[integrate_titles['show_type'].isna().index,integrate_titles.columns.get_loc('show_type')] =  integrate_titles.iloc[integrate_titles['show_type'].isna().index,integrate_titles.columns.get_loc('type')]

integrate_titles.iloc[np.where(integrate_titles['date_added'].isna())[0],integrate_titles.columns.get_loc('date_added')] =  integrate_titles.iloc[np.where(integrate_titles['date_added'].isna())[0],integrate_titles.columns.get_loc('from')]

integrate_titles=integrate_titles.drop(columns=["type"], axis=1)
integrate_titles=integrate_titles.drop(columns=["from"], axis=1)

# contry e production_country sono ridondanti, possiamo buttarne via una


In [34]:
print(integrate_titles['date_added'].isna().sum())
print(integrate_titles['show_type'].isna().sum())
#

1
0


Infine, ci sono colonne che già ora possiamo giudicare inutili?

In [35]:
# ...completare qui ...

integrate_titles=integrate_titles.drop(columns=["Unnamed: 0"], axis=1)

integrate_titles.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4715 entries, 0 to 1069
Data columns (total 19 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    4715 non-null   object 
 1   title                 4715 non-null   object 
 2   release_year          4715 non-null   int64  
 3   age_certification     2857 non-null   object 
 4   runtime               4715 non-null   int64  
 5   genres                4715 non-null   object 
 6   production_countries  4715 non-null   object 
 7   seasons               1716 non-null   float64
 8   imdb_id               4224 non-null   object 
 9   imdb_score            4170 non-null   float64
 10  imdb_votes            4165 non-null   float64
 11  tmdb_popularity       4679 non-null   float64
 12  tmdb_score            4556 non-null   float64
 13  director              3020 non-null   object 
 14  country               4210 non-null   object 
 15  date_added            4714

### Carichiamo altre due tabelle, che forniscono fonti di informazioni aggiuntive

In [36]:
netflix_credits = pd.read_csv("netflix_credits.csv")
disney_credits = pd.read_csv("disney_credits.csv")

come sempre, stampiamo le prime righe di ciascun dataFrame

In [37]:
netflix_credits.head()


person_id       id                name                  character   role
0       3180  tm82169  Sylvester Stallone      Robert "Rocky" Balboa  ACTOR
1       7748  tm82169         Talia Shire  Adrianna "Adrian" Pennino  ACTOR
2       7749  tm82169          Burt Young             Paulie Pennino  ACTOR
3       7750  tm82169       Carl Weathers               Apollo Creed  ACTOR
4       7751  tm82169    Burgess Meredith            Mickey Goldmill  ACTOR

In [38]:
disney_credits.head()

person_id       id            name         character   role
0      35549  tm89464  Maureen O'Hara      Doris Walker  ACTOR
1      57832  tm89464      John Payne       Fred Gailey  ACTOR
2      57833  tm89464    Edmund Gwenn      Kris Kringle  ACTOR
3      25096  tm89464    Natalie Wood      Susan Walker  ACTOR
4      27185  tm89464     Porter Hall  Granville Sawyer  ACTOR

Possiamo osservare che le tabelle credits e titles, hanno informazioni complementari.  In particolare, credits riporta delle informazinoi aggiuntive riguardanti attori e carattere interpretato.

Come prima, procediamo con l'integrazione delle due tabelle, che questa volta è più facile (lo schema è proprio lo stesso!)

In [39]:
integrate_credits = pd.concat([netflix_credits,disney_credits])

Data la complementarietà delle informazioni che troviamo nelle cartelle titles e credits, possiamo decidere di fare un **join** tra le due tabelle, per includere tutte le informazioni in un unico dataset.


In [104]:
# ...completare qui usando la funzione merge per fare il join
final_table=pd.merge(integrate_titles,
         integrate_credits,
         how = "inner",
         left_on = "id",
         right_on = "id");

verifichiamo cosa sia successo guardando le dimensioni di ciascun dataframe

In [105]:
print(integrate_titles.shape)
print(integrate_credits.shape)
print(final_table.shape)

(4715, 19)
(110791, 5)
(66189, 23)


vi tornano le dimensioni?

In [106]:
final_table.head()

id  title  release_year age_certification  runtime  \
0  tm82169  Rocky          1976                PG      119   
1  tm82169  Rocky          1976                PG      119   
2  tm82169  Rocky          1976                PG      119   
3  tm82169  Rocky          1976                PG      119   
4  tm82169  Rocky          1976                PG      119   

               genres production_countries  seasons    imdb_id  imdb_score  \
0  ['drama', 'sport']               ['US']      NaN  tt0075148         8.1   
1  ['drama', 'sport']               ['US']      NaN  tt0075148         8.1   
2  ['drama', 'sport']               ['US']      NaN  tt0075148         8.1   
3  ['drama', 'sport']               ['US']      NaN  tt0075148         8.1   
4  ['drama', 'sport']               ['US']      NaN  tt0075148         8.1   

   ...          director        country      date_added rating duration  \
0  ...  John G. Avildsen  United States  August 1, 2019     PG  120 min   
1  ...  John G. Avildsen  United States  August 1, 2019     PG  120 min   
2  ...  John G. Avildsen  United States  August 1, 2019     PG  120 min   
3  ...  John G. Avildsen  United States  August 1, 2019     PG  120 min   
4  ...  John G. Avildsen  United States  August 1, 2019     PG  120 min   

  show_type person_id                name                  character   role  
0     MOVIE      3180  Sylvester Stallone      Robert "Rocky" Balboa  ACTOR  
1     MOVIE      7748         Talia Shire  Adrianna "Adrian" Pennino  ACTOR  
2     MOVIE      7749          Burt Young             Paulie Pennino  ACTOR  
3     MOVIE      7750       Carl Weathers               Apollo Creed  ACTOR  
4     MOVIE      7751    Burgess Meredith            Mickey Goldmill  ACTOR  

[5 rows x 23 columns]

In [107]:
print(final_table.columns)

Index(['id', 'title', 'release_year', 'age_certification', 'runtime', 'genres',
       'production_countries', 'seasons', 'imdb_id', 'imdb_score',
       'imdb_votes', 'tmdb_popularity', 'tmdb_score', 'director', 'country',
       'date_added', 'rating', 'duration', 'show_type', 'person_id', 'name',
       'character', 'role'],
      dtype='object')


e visualizziamo gli eventuali elementi mancanti

In [108]:
final_table.isnull().sum()

id                          0
title                       0
release_year                0
age_certification       22496
runtime                     0
genres                      0
production_countries        0
seasons                 54660
imdb_id                  3968
imdb_score               4251
imdb_votes               4288
tmdb_popularity             0
tmdb_score                525
director                11698
country                  4063
date_added                  1
rating                     30
duration                    2
show_type                   0
person_id                   0
name                        0
character                7059
role                        0
dtype: int64

concentriamoci sulla colonna date_added, intanto visualizzandola:

In [109]:
print(final_table['date_added'])

0            August 1, 2019
1            August 1, 2019
2            August 1, 2019
3            August 1, 2019
4            August 1, 2019
                ...        
66184      October 20, 2021
66185    September 15, 2021
66186          May 28, 2021
66187          May 28, 2021
66188      October 29, 2021
Name: date_added, Length: 66189, dtype: object


vediamo che giorni mesi e anni sono presenti. Potrebbe essere utile, però, separare queste informazioni, incorporandole in tre colonne diverse.

Facciamolo insieme, con due diversi passi:

1. dobbiamo trasformare questa colonna in formato 'datetime' utilizzando la funzione di pandas 'to_datetime'

In [111]:
final_table['date_added'] = pd.to_datetime(final_table['date_added'])

In [112]:
print(final_table['date_added'])

0       2019-08-01
1       2019-08-01
2       2019-08-01
3       2019-08-01
4       2019-08-01
           ...    
66184   2021-10-20
66185   2021-09-15
66186   2021-05-28
66187   2021-05-28
66188   2021-10-29
Name: date_added, Length: 66189, dtype: datetime64[ns]


poi, possiamo separare le varie entrate in tre nuove colonne, che si chiameranno day, month e year.

In [113]:
final_table['year'] = final_table['date_added'].dt.year
final_table['month'] = final_table['date_added'].dt.month
final_table['day'] = final_table['date_added'].dt.day

ora visualizziamo il risultato

In [114]:
final_table.head()

id  title  release_year age_certification  runtime  \
0  tm82169  Rocky          1976                PG      119   
1  tm82169  Rocky          1976                PG      119   
2  tm82169  Rocky          1976                PG      119   
3  tm82169  Rocky          1976                PG      119   
4  tm82169  Rocky          1976                PG      119   

               genres production_countries  seasons    imdb_id  imdb_score  \
0  ['drama', 'sport']               ['US']      NaN  tt0075148         8.1   
1  ['drama', 'sport']               ['US']      NaN  tt0075148         8.1   
2  ['drama', 'sport']               ['US']      NaN  tt0075148         8.1   
3  ['drama', 'sport']               ['US']      NaN  tt0075148         8.1   
4  ['drama', 'sport']               ['US']      NaN  tt0075148         8.1   

   ...  rating  duration  show_type person_id                name  \
0  ...      PG   120 min      MOVIE      3180  Sylvester Stallone   
1  ...      PG   120 min      MOVIE      7748         Talia Shire   
2  ...      PG   120 min      MOVIE      7749          Burt Young   
3  ...      PG   120 min      MOVIE      7750       Carl Weathers   
4  ...      PG   120 min      MOVIE      7751    Burgess Meredith   

                   character   role    year month  day  
0      Robert "Rocky" Balboa  ACTOR  2019.0   8.0  1.0  
1  Adrianna "Adrian" Pennino  ACTOR  2019.0   8.0  1.0  
2             Paulie Pennino  ACTOR  2019.0   8.0  1.0  
3               Apollo Creed  ACTOR  2019.0   8.0  1.0  
4            Mickey Goldmill  ACTOR  2019.0   8.0  1.0  

[5 rows x 26 columns]

ora creiamo una copia della tabella, e modifichiamo la durata dei film da minuti a ore

Attenzione, ci sono dei campioni in cui la durata è in season, ma il tipo è MOVIE.
Questi possono essere eliminati

In [183]:
final_table=pd.merge(integrate_titles,
         integrate_credits,
         how = "inner",
         left_on = "id",
         right_on = "id");
movies_indexes = np.where(final_table['show_type']=='MOVIE')[0]
#troviamo questi campioni ed eliminiamoli
print(np.sum(final_table.iloc[movies_indexes,final_table.columns.get_loc('duration')].str.contains('Season')))
final_table = final_table.drop(final_table[final_table['duration'].str.contains('Season', na=False)].index)
movies_indexes = np.where(final_table['show_type']=='MOVIE')[0]
print(np.sum(final_table.iloc[movies_indexes,final_table.columns.get_loc('duration')].str.contains('Season')))

406
0


a questo punto, possiamo modificare la durata in ore, per prima cosa eliminiamo la parola minuti

In [184]:
final_table['duration']=final_table['duration'].replace(' min','',regex=True)

quindi dividere per 60 (dopo conversione float)

In [185]:
print(final_table.iloc[movies_indexes,final_table.columns.get_loc('duration')])

0        120
1        120
2        120
3        120
4        120
        ... 
66178     41
66179     41
66180     41
66181     41
66188     44
Name: duration, Length: 54254, dtype: object


e finalmente, aggiungere un suffisso per indicare le ore, per esempio 'hours', tornando a stringa

In [186]:
final_table.iloc[movies_indexes,final_table.columns.get_loc('duration')] = final_table.iloc[movies_indexes,final_table.columns.get_loc('duration')].astype(float)/60
final_table.iloc[movies_indexes,final_table.columns.get_loc('duration')] = final_table.iloc[movies_indexes,final_table.columns.get_loc('duration')].astype(str).str[:5]+ ' hours'

stampiamo il risultato finale

In [187]:
print(final_table['duration'])

0          2.0 hours
1          2.0 hours
2          2.0 hours
3          2.0 hours
4          2.0 hours
            ...     
66178    0.683 hours
66179    0.683 hours
66180    0.683 hours
66181    0.683 hours
66188    0.733 hours
Name: duration, Length: 54516, dtype: object
